[red-knot] Add diagnostic for class-object access to pure instance variables and class literal repr in diagnostics

mishamsk opened the pull request:
## Summary

This is at best a band-aid solution for #15963 . See my first comment below for discussion.

As of time of writing:

* Fixes unbound class attribute access. Doesn't return a bound symbol anymore and thus causes a `[unresolved-attribute]` diagnostic
* Intentionally as a separate commit:
  * Changes`Type::ClassLiteral` representation to `type[ClassName]` matching mypy (this was already the case for `Type::SubclassOf`)
  * Changes representation of a union of 2+ types to `types.UnionType[TypeOne, TypeTwo, ...]` also matching mypy

I didn't not add a separate new lint, despite the issue description. I am not sure why `[unresolved-attribute]` is not appropriate. Unbound attributes indeed do not exist at runtime, so the language seems perfectly appropriate.

## Test Plan

cargo nextest run -p red_knot_python_semantic --no-fail-fast

mishamsk commented:
@sharkdp first, sorry for taking so long, I got distracted by work ;-) 

second, sorry for a long read for such a small thing. However, since we can't just talk it over - I'll dump why I think it is a band aid and PR marked as draft. 

Disclaimer: by no means I am implying there is anything wrong with knot code. I totally understand that you are moving crazy fast and this is one huge undertaking, so I totally get it when something is not "perfect" or a temporary solution.

The reason it took me multiple hours to understand what's happening is that I got confused by the fact that knot uses a notion of "Symbol" for both real symbols and unbound declarations, which are not symbols at runtime. They only end up in annotations:

```
In [1]: class C:
   ...:     not_a_symbol: int
   ...:

In [2]: dir(C)
Out[2]:
['__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__']

In [3]: C.__annotations__
Out[3]: {'not_a_symbol': int}
```

Even though I quickly found the passage rgd `Declaration` that is used for both types of statements, the fact that they both create a symbol wasn't readily apparent.

Further, confusing the matter was the fact that `red_knot_python_semantic::symbol::Symbol` (not entirely helpfully matching the `red_knot_python_semantic::semantic_index::symbol::Symbol`) returns the enum `Boundness` for both types of "symbols". So pure (unbound) declarations will be `Symbol::Type(ty, Boundness::Bound)`.

## Ideas

### Symbol::Type
Tbh and imho symbols and declarations should be split. At least in this context. But I felt like doing so would cause too many changes in too many places. Something I would only do if you confirm such an idea. There are multiple ways this can be done:

* Very invasive - create something separate from symbols for declarations and match them to bindings / use sites when necessary. Probably too much of a change and disruption 
* Extend `Boundness` and add two more values: `Declared` and `PossiblyDeclared`
* In the same vain - extend `Symbol::Type` tuple by adding some enum with the sub-kind of the symbol. This won't solve the confusing "Boundness::Bound unbound declared symbol" thing though
* Maybe there is even better way

### Split `symbol` function

`red_knot_python_semantic::types::symbol` function is used in multiple places to get the symbol type & boundness. But not all of them, e.g. `Class::own_instance_member` does it's own but similar thing. In the same vain, it was confusing me a bit. I'd split it or at least extract a more narrower common part out of it. 

E.g. why would we need to check for `TYPE_CHECKING` on every invocation, if the only place where it may be triggered from is `known_module_symbol` (which calls `symbol` through `global_symbol`) and inversely why looking up from `global_symbol` would need to check for dunder `__slots__`.

Again, really sorry for a long dump, especially if there are very good reasons for all of this that I just failed to identify.

mishamsk commented:
well. here is one false-positive from this change: https://github.com/astral-sh/ruff/actions/runs/13212332102/job/36887424507?pr=16036#step:7:94

ClassVar in stubs/ABCs needs to be handled as bound... I'll look into it later

sharkdp commented:
As mentioned on Discord, I'll split my reply into a few comments

> * Intentionally as a separate commit:
>   * Changes`Type::ClassLiteral` representation to `type[ClassName]` matching mypy (this was already the case for `Type::SubclassOf`)
>   * Changes representation of a union of 2+ types to `types.UnionType[TypeOne, TypeTwo, ...]` also matching mypy

Commenting on this first. This should definitely be discussed elsewhere (ideally as a ticket first), but I don't think that we want either of these changes. 

Unlike mypy, we deliberately distinguish `Literal[C]` from `type[C]`, as they denote different types. We don't know yet if we really want to use `Literal[C]` as the spelling of that type, but I don't think we want to mix up `Literal[C]` and `type[C]`.

The difference between the two is:

* `Literal[C]` is the type of the class object `C`, for a `class C: …`. It is a singleton type with one single inhabitant: the class object `C`.
* `type[C]` is the type of all class objects `S`, where `S` is a subclass of `C`, or `C` itself.

More concretely, both `C` and `D` are of type `type[C]`, but only `C` is of type `Literal[C]`:

```py
class C: ...
class D(C): ...
```


The union-type representation change would be more stylistic, but I don't see any argument for choosing `types.UnionType[TypeOne, TypeTwo]` over `TypeOne | TypeTwo`?

mishamsk commented:
> but I don't think that we want either of these changes

Yeah, that's why put them in a separate commit. I had a hunch ;-) I just saw comments like "we want to improve the message". I'll drop this commit, no prob.

> The difference between the two

yes, I do understand that and I saw that you use `type[]` only for `Type::SubclassOf`. I just thought that `Literal` is a bit unconventional coming from mypy/pyright. If you are planning to stray far from conventions, maybe changing the outer message would be an interesting approach. Instead of 
```
Type `Literal[A]` has no attribute `non_existent`
```

generate 
```
Class `A` has no attribute `non_existent`
```

> The union-type representation change would be more stylistic, but I don't see any argument for choosing types.UnionType[TypeOne, TypeTwo] over TypeOne | TypeTwo?

same here. I went "conventional" way (mypy). If you keep `Literal`, it would be `Literal[TypeOne, TypeTwo]` I guess

sharkdp commented:
> I just saw comments like "we want to improve the message".

> Class `A` has no attribute `non_existent`

I should have described this better in the TODO comment:

https://github.com/astral-sh/ruff/blob/f7819e553f0dbb22afcf31ea77938d529fe52818/crates/red_knot_python_semantic/resources/mdtest/attributes.md?plain=1#L53-L56

I didn't specifically want to talk about the `Literal[C]` part here, but I agree that this could *also* be improved. I was rather thinking about the following: This specific diagnostic here is raised if someone tries to access a *pure instance variable* on the class object (not for a non-existent attribute). So I was thinking of giving them a hint and explaining that in more detail: "The attribute `inferred_from_value` can only be accessed on instances of class `C`, but not on the class object itself."

sharkdp commented:
I now had a more detailed look at the ideas you proposed and they do look very interesting to me. They definitely go in the right direction. What I would prefer is if we could decouple any refactoring from this functional change here. Your "bandaid" solution doesn't look too bad to me, given our current API restrictions.

It would be great if you could remove the second commit from this branch, then we could focus on the functional change here and try to understand the errors in stub files better.

mishamsk commented:
@sharkdp first rdg the PR:

- reverted class literal display code
- added ABC, ABCMeta to known classes
- added handling of stubs & classvars to `own_class_member` with some tests
  - had to extract the code from `symbol` as I needed to get qualifiers, not just type... this is possibly a performance hit, since `symbol` used salsa query internally... something had to give (either big changes across the board, or this). To my defense `own_instance_member` is also re-creating the logic instead of using a shared query

Am I right, that you also want me to create a separate diagnostic for this specific access patter with the message you suggested? Didn't have time for this yet.

Also, I didn't do extensive validation that ABC's are properly caught in all possible scenarios. But at least based on two tests they do.

Now back to the general design. Or should I move this to #14297? 

I saw #14297 and related comments, but not https://github.com/astral-sh/ruff/blob/main/crates/red_knot_python_semantic/resources/mdtest/boundness_declaredness/public.md - thanks for the link. I think the table is awesome, it literally suggest the perfect way to model this.

I'd refactor symbol to:
```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum BindingStatus<'db> {
    Bound(Type<'db>),
    PossiblyUnbound(Type<'db>),
    Unbound,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum DeclarationStatus<'db> {
    Declared(TypeAndQualifiers<'db>),
    PossiblyDeclared(TypeAndQualifiers<'db>),
    Undeclared,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) enum Symbol<'db> {
    Known(BindingStatus<'db>, DeclarationStatus<'db>),
    Unknown,
}
```

return this from `symbol` query and then let the caller apply whatever logic is appropriate. `Symbol` can have methods that resolve types between declarations & bindings based on your table from that file too.

sharkdp commented on crates/red_knot_python_semantic/src/types.rs:
````diff
@@ -4140,8 +4167,123 @@ impl<'db> Class<'db> {
+            // class FromCExtension:
+            //    symbol: int
+            // ```
+            if !is_stub && !is_abstract && !is_bound {
````
Instead of designing a special case for stubs and abstract base classes, shouldn't we design a special case for `ClassVar`s? The false positive you were getting from stubs was that `Literal[timezone]` has no attribute `utc`. But `utc` is explicitly [annotated as a `ClassVar`](https://github.com/python/typeshed/blob/24c78b9e0dff457275092025a14387dac206f5d6/stdlib/datetime.pyi#L29).

The idea of [the ticket](https://github.com/astral-sh/ruff/issues/15963) was to detect class-object access to *instance variables*. If we see a `utc: ClassVar[…]` annotation (in a stub or not), we should treat that as a pure class variable instead and _allow_ access on class objects.

It would be great if we would not have to special-case stubs and ABCs.

mishamsk replied:
sorry. got carried away I guess... this is fixed. I'll list the full list of changes in a new thread below

mishamsk commented:
- remove ABC detection
- remove special casing for class member access in stubs
- improve error message for invalid pure instance attr access from class in load context 
- added error message for invalid pure instance attr access from class in store context 
- switched the symbol lookup implementation inside class member function into salsa query, similar to the not-so-shared `symbol` function

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3482,22 +3496,39 @@ impl<'db> TypeInferenceBuilder<'db> {
+                            &INVALID_ATTRIBUTE_ACCESS,
+                            attribute.into(),
+                            format_args!(
+                                "Cannot assign to pure instance attribute `{attr}` from a class of type `{ty}`",
```
I think we can omit the word "pure" here -- assigning to _any_ instance attribute from the class object is illegal

```suggestion
                                "Cannot assign to instance attribute `{attr}` from the class object `{ty}`",
```

mishamsk replied:
THe word "pure" comes from @sharkdp and afaik the intended target of this PR/issue. 

Do you mean that in red_knot you want to restrict this as well:

```python
class C:
    attr: str = "class_value"

    def __init__(self):
        self.attr = "instance value"

C.attr = "other class value"
```

if so, I think it is a deviation from what the issue said, so want to double-check that's the intent here.

AlexWaygood replied:
> Do you mean that in red_knot you want to restrict this as well:

Yes. We should allow instance attributes like that to be _accessed_ from the class object but not to be _assigned_ on the class object

AlexWaygood replied:
And for "pure" instance attributes -- those which do not have a default value set in the class body -- they should be neither readable nor writable on the class object :-)

mishamsk replied:
@AlexWaygood I applied naming changes per your suggestions. 

I did not implement the new case:

```python
class C:
    attr: str = "class_value"

    def __init__(self):
        self.attr = "instance value"

C.attr = "other class value"
```

for two reasons:

- outside of the scope of original issue
- more importantly: this would need changing member or onw_class_member API to return qualifiers. I think this will collide with what @sharkdp is doing with descriptors now. It would be much easier to add this one last case to tests & emit diagnostic when API's are changed

carljm replied:
> Yes. We should allow instance attributes like that to be _accessed_ from the class object but not to be _assigned_ on the class object. On the class object, they should be readable but not writable

What's the rationale for this? It's legal at runtime, both pyright and mypy allow it, I'm not sure I see why it must be disallowed.

mishamsk replied:
btw. I didn't argue with @AlexWaygood but I am with @carljm - IMHO there is no reason to forbid that. It was not and is not apparent to me

AlexWaygood replied:
> What's the rationale for this? It's legal at runtime, both pyright and mypy allow it, I'm not sure I see why it must be disallowed.

It's something I'd want a type checker to flag if I'm running a type checker on my Python code. Sure, it's legal at runtime, but it seems like it's probably a user mistake to me if the variable has been explicitly declared as an instance attribute but it's being modified on the class object rather than an instance. Having said all this, your pushback makes me think that the error should probably be disabled by default, since you're right that it isn't really _unsafe_ for an instance attribute with a class-level default to be assigned on the class object.

mishamsk replied:
As of current PR it will not emit a diagnostic. Do you suggest adding a configuration and implement it under a flag? Seems like a reasonable compromise. 

I haven't researched knot configuration capabilities yet. Maybe a follow up?

AlexWaygood replied:
@mishamsk you don't need to worry about it for now: it's not implemented on `main`, and it shouldn't be implemented as part of this PR :-) I was just responding to @carljm's (very reasonable) question about my earlier assertion. It could absolutely be a followup, but first we need to reach agreement on whether it's something we do indeed want ;-) and it might not be a priority for us to implement in any case

AlexWaygood reviewed (commented):
Nice! I haven't done a full review, but I spotted a few things looking over quickly

mishamsk commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -4144,14 +4144,100 @@ impl<'db> Class<'db> {
     pub(crate) fn own_class_member(self, db: &'db dyn Db, name: &str) -> Symbol<'db> {
-        let scope = self.body_scope(db);
-        symbol(db, scope, name)
+        #[salsa::tracked]
+        fn class_symbol_by_id<'db>(
+            db: &'db dyn Db,
+            scope: ScopeId<'db>,
+            is_dunder_slots: bool,
+            symbol_id: ScopedSymbolId,
+        ) -> Symbol<'db> {
```
thanks. tbh I am not sure there is a point in query at all. My first version didn't have a query, and the codsped showed the same hit to performance. I intentionally matched the existing `symbol` function and it's internal query syntax to simplify merging/unifying them later.

I can apply your suggestion, but I'd just drop query. LMK

MichaReiser replied:
I don't think we can drop the query. The query is necessary for better incremental computation because `symbol_from_declarations` calls into the visibility constraint evaluation that calls `node_ref` on some nodes that don't necessarily belong to the same module as the module calling `Type::member`.

AlexWaygood replied:
> tbh I am not sure there is a point in query at all.

I'm not totally sure TBH. It's important to use queries when we have code that might be "reaching across the module boundary" -- this helps make sure that red-knot is incremental in its type-checking, and only recomputes the things it absolutely needs to when something changes in a file. I'm not confident on whether that applies here or not. @MichaReiser is our expert on these matters.

AlexWaygood replied:
Ah and @MichaReiser got there before I even finished writing that message 😄

mishamsk replied:
@MichaReiser thanks, very helpful explanation. I think I now understand how you choose where to use query. At least the case of cross-boundary vs. same-file analysis.

I'll apply the suggested change here, but just FYI - `own_instance_member` should be changed too. Going back to my initial suggestion to refactor the symbol business (most recent [comment here](https://github.com/astral-sh/ruff/pull/16036#issuecomment-2649552081)). Not as part of this PR, but I'd be happy to do that if the team allows.

mishamsk replied:
fixed

carljm replied:
https://github.com/astral-sh/ruff/issues/16172 to track the `own_instance_member` issue

mishamsk commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3455,15 +3455,29 @@ impl<'db> TypeInferenceBuilder<'db> {
-                        value_ty.display(self.db()),
-                        attr.id
+                match value_ty {
+                    Type::ClassLiteral(_) | Type::SubclassOf(_) => {
+                        self.context.report_lint(
+                            &UNRESOLVED_ATTRIBUTE,
+                            attribute.into(),
+                            format_args!(
+                                "The attribute `{}` can only be accessed on instances of type `{}`, but not on the class object itself.",
+                                attr.id,
+                                value_ty.display(self.db())
+                            ),
+                        );
+                    }
```
The idea was to get a nice helpful message in a common case of accidentally trying to assign existing instance var via class. But you are right, I didn't think about entirely non-existent attrs.

I suggest to simplify/unify the message with the Instance arm above. To make a nice message that will distinguish this two cases would lead to doing a second symbol search through instances, probably unreasonable overhead.

@sharkdp are you ok with that?

AlexWaygood replied:
Helpful error messages are important when it comes to creating a tool that users enjoy running on their code! And performance doesn't matter quite so much in this specific location. Here we already know we're about to emit a diagnostic, so we only pay the cost of doing something slow here if there's a typing error in the user's code. That's hopefully pretty rare; most of the time most users' code won't have typing errors in it.

sharkdp replied:
> To make a nice message that will distinguish this two cases would lead to doing a second symbol search through instances, probably unreasonable overhead.

I think I would be fine with a possible performance hit for this, especially if we can perform that lookup only if we would emit a diagnostic anyway, i.e. if we would not have to pay a cost for this if everything is fine.

mishamsk replied:
ok. we'll do later today

mishamsk replied:
this should be fixed now

mishamsk commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -4144,14 +4144,100 @@ impl<'db> Class<'db> {
+                            // if the symbol is PossibleUnobund. This requires a bit of
+                            // design work though as we might want a different behavior for stubs and for
+                            // normal modules.
+                            Symbol::Type(declared_ty, Boundness::Bound)
```
ah. that's new API if I am not mistaken. Thanks for the tip

carljm commented on crates/red_knot_python_semantic/resources/mdtest/attributes.md:
```diff
@@ -50,13 +50,12 @@ c_instance.inferred_from_param = None
-# TODO: we already show an error here but the message might be improved?
 # mypy shows no error here, but pyright raises "reportAttributeAccessIssue"
-# error: [unresolved-attribute] "Type `Literal[C]` has no attribute `inferred_from_value`"
+# error: [unresolved-attribute] "Attribute `inferred_from_value` can only be accessed on instances of type `Literal[C]`, not on the class object itself."
```
Since we are talking about the instance type, I think it would be less confusing if this error message used the instance type name (aka just the class name):

```suggestion
# error: [unresolved-attribute] "Attribute `inferred_from_value` can only be accessed on instances of type `C`, not on the class object itself."
```

mishamsk replied:
@carljm I totally agree. But this is a shared display logic. I originally included another change in this PR related to displaying classes & unions, which touched a lot of mdtest, but reverted it per @sharkdp request. Now I am hesitant to include anything like that in this PR. I doubt that special-casing for this single instance would be welcomed either => I'd rather leave message improvements to another PR and look at them more holistically

sharkdp replied:
> but reverted it per @sharkdp request.

This previous change affected how we displayed class literal types in general (`type[C]` instead of `Literal[C]`). But here, we don't want to change how class literal types are displayed, we rather want some some special handling that turns the `Type::ClassLiteral(class)` type into a `Type::Instance(class)` type before displaying. I am not sure if that is an easy change, though, since we would also need to handle unions etc.

If it turns out to be more complicated, we could also work around this for now by rephrasing that sentence a bit, e.g. "can only be accessed on instances, not on the class object `Literal[C]` itself."?

carljm replied:
Tweaking the wording of a single diagnostic message, which is newly introduced in this PR, is quite different from wholesale changes to the display of some kinds of types. I am suggesting the former, not the latter.

I think the change I suggested is pretty easy to implement (and doesn't even require a general type transform using `to_instance()`) because it is already the case that we only emit this specialized diagnostic if the receiver type is exactly a `ClassLiteral` or `SubclassOf` type (not e.g. a union). So we have access to the `class` and could just use its name.

But that said, I like @sharkdp 's suggestion even better, to just reword the message so we mention the actual type where we are discussing "the class object itself" rather than where we are discussing instances. This rewording makes sense, because it is the class type, not the instance type, that we have at hand. It's also the way the "assignment" version of this same diagnostic is already worded.

mishamsk replied:
applied the suggested wording

carljm commented on crates/red_knot_python_semantic/resources/mdtest/expression/attribute.md:
```diff
@@ -29,6 +29,6 @@ def _(flag: bool):
     reveal_type(A.possibly_unbound)  # revealed: str
 
-    # error: [unresolved-attribute] "Type `Literal[A]` has no attribute `non_existent`"
+    # error: [unresolved-attribute] "Attribute `non_existent` can only be accessed on instances of type `Literal[A]`, not on the class object itself."
```
This error message change looks like a regression that should be reversed. The new error message is misleading because it implies that accessing `non_existent` on the instance type would work, when it would not, because the attribute doesn't exist at all. (I think maybe this is also discussed in another comment thread below?)

carljm replied:
Oops ignore, commented before you made the update.

carljm commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -4144,14 +4144,98 @@ impl<'db> Class<'db> {
-        let scope = self.body_scope(db);
-        symbol(db, scope, name)
+        #[salsa::tracked]
+        fn class_symbol_by_id<'db>(
```
How different is this from `symbol_by_id`? It looks like a lot of it is copy-pasted. I'm a bit hesitant to copy-paste such a significant chunk of code, with known TODOs and all. (I realize `own_instance_member` also implements its own lookup logic, but I think that logic is much more significantly different from `symbol_by_id`.

For context, `symbol_by_id` used to be a standalone function before it was moved inside `symbol`. And it already has multiple arguments, so already pays the cost of Salsa argument interning. I don't see much reason not to move it back outside of `symbol`, and add an argument to it would allow it to serve this use case, too. Or an interesting alternative approach would be to have it return a `SymbolAndQualifiers` which might allow the logic here to be implemented externally?

At the very least, it would be easier to review and understand the differences in behavior!

mishamsk replied:
Wholeheartedly agree. I suggested refactoring the whole symbol thing. I also think it should return `SymbolAndQualifiers` and that the `Symbol` enum should be changed to push the logic of combining declared and inferred types to the caller. Does everyone want me to do this in this PR though? @AlexWaygood @sharkdp

carljm replied:
Yes, I think it would make sense to pursue that refactor in this PR; I don't think there is any particular urgency to quickly land a copy-paste version of the functionality. I think it's safe to interpret the thumbs-ups above as team agreement :)

mishamsk replied:
Ok. I have a short holiday this weekend, we'll get back to it Monday

mishamsk replied:
@carljm I spent some time today, trying to find a suitable unification between 3 different symbol search queries, and haven't landed on a design yet. I am not sure it makes sense, but I'll do another pass tomorrow before a final verdict.

One issue with keeping this refractor and #16172 in scope of this PR is that main is moving forward quite fast. E.g. some of the code I've used moved to symbol.rs mod as a private function after #16152 and the symbol implementation itself was changed.

Not sure how many other conflicting PRs are in the pipe, I'll do my best to quickly create my version of refactoring in hopes that no new divergence occurs.

mishamsk replied:
@carljm I've unified the two queries. And then made `symbol_by_id` a regular function in a separate commit (easy to revert) after reading #16268 and given that `symbol_by_id` now had to return a bigger structure and thus query performance would be reduced.

tbh. it feels a bit wrong and suboptimal due to the whole Boundness not always meaning boundness. But since redesigning of this is out of scope and you asked me to unify the two logics - I did what I think is the best intermediary solutions.

If you think it looks good, I would re-request review from the wider group

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3455,15 +3456,56 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                        value_ty.display(self.db())
+                                    ),
+                                );
+                        } else {
+                            self.context.report_lint(
+                                &UNRESOLVED_ATTRIBUTE,
+                                attribute.into(),
+                                format_args!(
+                                    "Type `{}` has no attribute `{}`",
+                                    value_ty.display(self.db()),
+                                    attr.id
+                                ),
+                            );
+                        }
+                    }
+                    _ => self.context.report_lint(
+                        &UNRESOLVED_ATTRIBUTE,
+                        attribute.into(),
+                        format_args!(
+                            "Type `{}` has no attribute `{}`",
+                            value_ty.display(self.db()),
+                            attr.id
+                        ),
```
I think it would be pretty easy to refactor this to avoid duplicating this default error message. Put the entire `match value_ty` inside `let bound_on_instance = match value_ty { ... }` where the fallback case assigns `false`, and then follow that with `if bound_on_instance { ... <access on instance diagnostic> ... } else { ... <default diagnostic> ... }`

mishamsk replied:
yes, indeed, thanks. applied this and merged with changes from #16160

carljm reviewed (commented):
Thank you!

carljm commented:
I suspect the performance regression on the incremental benchmark here is not avoidable. We just need more cached query results now that we distinguish more ways to resolve symbols, and more cached query results means more re-validation cost in Salsa.

AlexWaygood commented:
> I suspect the performance regression on the incremental benchmark here is not avoidable. We just need more cached query results now that we distinguish more ways to resolve symbols, and more cached query results means more re-validation cost in Salsa.

Note that we haven't got up to date benchmark numbers on this PR right now, as the assertion in the benchmark regarding the number of diagnostics has been failing for the last few commits, so codspeed has not been running

carljm commented:
> I suspect the performance regression on the incremental benchmark here is not avoidable. We just need more cached query results now that we distinguish more ways to resolve symbols, and more cached query results means more re-validation cost in Salsa.

I'm also questioning this conclusion -- I think if we were able to make `symbol` return a `SymbolAndQualifiers` and that were enough to let us strip out classvars externally, that might address the perf impact here, since it would mean we could do this without increasing the number of cached memos (instead we'd just be slightly increasing the data stored in each one, which is not as expensive in Salsa incremental re-validation).

MichaReiser commented:
I haven't reviewed this PR yet because there are already a lot of folks involved. Feel free to ping me if or when a review from my side would be good.

mishamsk commented on crates/red_knot_python_semantic/resources/mdtest/attributes.md:
```diff
@@ -595,6 +590,9 @@ C.class_method()
 reveal_type(C.pure_class_variable)  # revealed: Unknown
 
+# TODO: should be no error when descriptor protocol is supported
+# and the assignment is properly attrbiuted to the class method.
```
fixed

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3653,16 +3654,44 @@ impl<'db> TypeInferenceBuilder<'db> {
             .unwrap_with_diagnostic(|lookup_error| match lookup_error {
                 LookupError::Unbound => {
+                    let bound_on_instance = match value_type {
+                    Type::ClassLiteral(ClassLiteralType { class }) => {
+                        !class.instance_member(self.db(), attr).0.is_unbound()
+                    }
+                    Type::SubclassOf(subclass_of @ SubclassOfType { .. }) => {
+                        match subclass_of.subclass_of() {
+                            ClassBase::Class(class) => {
+                                !class.instance_member(self.db(), attr).0.is_unbound()
+                            }
+                            ClassBase::Dynamic(_) => false,
```
I think we could even be a little stricter here and mark this branch `unreachable!`? Because `type[Any]` has all attributes, so we can never get `Unbound` back from getting a member off it.

The main advantage would be that in some sense neither `true` nor `false` seems right here, so it might make the code a bit clearer, and it would mean we fail fast if we violate this invariant. The disadvantage would be that if we somehow ship a very subtle violation of the invariant, it could cause a panic for a user.

I don't feel strongly about using `unreachable!`, open to whatever you prefer. But if we don't use `unreachable!`, I feel like `true` makes slightly more sense here than `false`, since `Any` also has all members.

mishamsk replied:
checked the codebase, seems like unreachable is being used in a ocuple of other places and also verified this invariant (mostly for myself). I think it is reasonable to use unreachable here. Some tests will fail quickly if it stops holding true.

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3653,16 +3654,44 @@ impl<'db> TypeInferenceBuilder<'db> {
+                    let bound_on_instance = match value_type {
+                    Type::ClassLiteral(ClassLiteralType { class }) => {
+                        !class.instance_member(self.db(), attr).0.is_unbound()
+                    }
+                    Type::SubclassOf(subclass_of @ SubclassOfType { .. }) => {
+                        match subclass_of.subclass_of() {
+                            ClassBase::Class(class) => {
+                                !class.instance_member(self.db(), attr).0.is_unbound()
+                            }
+                            ClassBase::Dynamic(_) => false,
+                        }
+                    }
+                    _ => false,
+                };
+
+                if bound_on_instance {
+                    self.context.report_lint(
+                            &UNRESOLVED_ATTRIBUTE,
+                            attribute,
+                            format_args!(
+                                "Attribute `{}` can only be accessed on instances, not on the class object `{}` itself.",
+                                attr.id,
+                                value_type.display(self.db())
+                            ),
+                        );
+                } else {
                     self.context.report_lint(
                         &UNRESOLVED_ATTRIBUTE,
                         attribute,
                         format_args!(
                             "Type `{}` has no attribute `{}`",
-                            value_type.display(db),
+                            value_type.display(self.db()),
                             attr.id
                         ),
                     );
-                    Type::unknown()
+                }
+
+                Type::unknown()
```
This all looks under-indented by one level? I think rustfmt struggles with the size of `infer.rs`, we need to break it up :/

mishamsk replied:
yeah, unfortunately rustfmt is not as good as ruff formatter for python! ;-)

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3692,22 +3721,54 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                Type::SubclassOf(subclass_of @ SubclassOfType { .. }) => {
+                                    match subclass_of.subclass_of() {
+                                        ClassBase::Class(class) => Some(class),
+                                        ClassBase::Dynamic(_) => None,
```
If we use `unreachable!` above, we could do so here as well.

carljm commented on crates/red_knot_python_semantic/src/symbol.rs:
```diff
@@ -362,92 +395,97 @@ impl<'db> From<Symbol<'db>> for SymbolAndQualifiers<'db> {
 
-/// Implementation of [`symbol`].
-fn symbol_impl<'db>(
+fn symbol_by_id<'db>(
```
It looks like we lost the `#[salsa::tracked]` annotation on this function. It's possible that we don't need it; we should see what the benchmarks say. If it is still a win to track this function, it also requires adding some trait derives on `Symbol` and `SymbolAndQualifiers`:

```diff
diff --git a/crates/red_knot_python_semantic/src/symbol.rs b/crates/red_knot_python_semantic/src/symbol.rs
index 63fd67203..d141c230a 100644
--- a/crates/red_knot_python_semantic/src/symbol.rs
+++ b/crates/red_knot_python_semantic/src/symbol.rs
@@ -37,7 +37,7 @@ pub(crate) enum Boundness {
 /// possibly_unbound:  Symbol::Type(Type::IntLiteral(2), Boundness::PossiblyUnbound),
 /// non_existent:      Symbol::Unbound,
 /// ```
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, salsa::Update)]
 pub(crate) enum Symbol<'db> {
     Type(Type<'db>, Boundness),
     Unbound,
@@ -365,7 +365,7 @@ pub(crate) type SymbolFromDeclarationsResult<'db> =
 /// that this comes with a [`CLASS_VAR`] type qualifier.
 ///
 /// [`CLASS_VAR`]: crate::types::TypeQualifiers::CLASS_VAR
-#[derive(Debug)]
+#[derive(Clone, Debug, Eq, PartialEq, salsa::Update)]
 pub(crate) struct SymbolAndQualifiers<'db>(pub(crate) Symbol<'db>, pub(crate) TypeQualifiers);

 impl SymbolAndQualifiers<'_> {
@@ -395,6 +395,7 @@ impl<'db> From<Symbol<'db>> for SymbolAndQualifiers<'db> {
     }
 }

+#[salsa::tracked]
 fn symbol_by_id<'db>(
     db: &'db dyn Db,
     scope: ScopeId<'db>,
```

mishamsk replied:
see comment below, closing this thread

carljm commented on crates/red_knot_python_semantic/src/symbol.rs:
```diff
@@ -183,6 +183,38 @@ pub(crate) fn symbol<'db>(db: &'db dyn Db, scope: ScopeId<'db>, name: &str) -> S
     symbol_impl(db, scope, name, RequiresExplicitReExport::No)
 }
 
+/// Infer the public type of a class symbol (its type as seen from outside its scope) in the given
```
We aren't getting any benchmark updates on this PR; the benchmark is failing to run because our assertion of the diagnostic output no longer matches. It looks like there is one new false positive on the line `TOMLDecodeError.__module__ = __name__`, where we think `__module__` is an instance attribute and therefore can't be set on the class. The attribute is defined in typeshed in `builtins.pyi` (on `object`) as just `__module__: str`. It seems to me like this should probably be defined in typeshed as a `ClassVar`? But maybe in a stub file we just need to consider all annotated assignments as "bound" even if they aren't (because the RHS is typically omitted in a stub file)? (This would dovetail with https://github.com/astral-sh/ruff/issues/16264 which is the same thing, but at module level instead of in class bodies; maybe we should just do it everywhere). Curious what you think about this, @AlexWaygood.

I'm OK with adding this false positive to the benchmark expected output for now and dealing with it as a separate follow-up. Here's the diff needed to get the benchmark running again (you can test locally with `cargo bench --bench red_knot`):

```diff
diff --git a/crates/ruff_benchmark/benches/red_knot.rs b/crates/ruff_benchmark/benches/red_knot.rs
index 11ec12577..beed5bd96 100644
--- a/crates/ruff_benchmark/benches/red_knot.rs
+++ b/crates/ruff_benchmark/benches/red_knot.rs
@@ -96,6 +96,13 @@ static EXPECTED_DIAGNOSTICS: &[KeyDiagnosticFields] = &[
         Cow::Borrowed("Unused blanket `type: ignore` directive"),
         Severity::Warning,
     ),
+    (
+        DiagnosticId::lint("invalid-attribute-access"),
+        Some("/src/tomllib/__init__.py"),
+        Some(270..296),
+        Cow::Borrowed("Cannot assign to instance attribute `__module__` from the class object `Literal[TOMLDecodeError]`"),
+        Severity::Error,
+    ),
 ];

 fn tomllib_path(file: &TestFile) -> SystemPathBuf {
```

mishamsk replied:
@carljm ran benches locally.

current version (no query):
![CleanShot 2025-02-21 at 15 07 45@2x](https://github.com/user-attachments/assets/5edfdd9f-56e5-476b-a047-6cd53e073ff9)

With query:
![CleanShot 2025-02-21 at 15 07 01@2x](https://github.com/user-attachments/assets/5e66eb2c-7805-4df6-abfd-535256f46078)

so I'll push non-query version as it seems either the same or slightly faster

carljm replied:
That seems fine to me. I think this is safe to do now that we have https://github.com/astral-sh/ruff/pull/16268. @MichaReiser do you have any reason that we should keep `symbol_by_id` tracked, if it doesn't seem to be a perf win to track it?

AlexWaygood replied:
I believe Micha experimented with removing the `#[tracked]` attribute on `symbol_by_id` in #16268, but Codspeed indicated that incremental performance was still much better if it was `#[tracked]`. Take a look at the individual commits in that PR and https://github.com/astral-sh/ruff/pull/16268#issuecomment-2672155862

carljm replied:
Hmm, it may be worth actually pushing up a commit that re-adds this and see what CodSpeed says (vs local run of the benchmark).

But it's also possible that perf characteristics have changed since Micha's experiment, due to other changes (or the changes in this PR.)

mishamsk replied:
you got it. just pushed. let's see what we get

sharkdp replied:
That would be consistent with what @MichaReiser and I found last week: Having `symbol_by_id` be a query helps with performance.

mishamsk replied:
so this one is good to be merged then? the current commit has `symbol_by_id` as a query

AlexWaygood replied:
yes, I think this thread can be resolved now :-)

carljm reviewed (commented):
This looks pretty good to me! I see your point about how it's sub-optimal until we address bound vs declared in `Symbol`, left an inline comment about that.

I think we need to get the benchmark running so we can see the perf impact of this, and evaluate the impact of tracking or not tracking `symbol_by_id`. But otherwise this looks good.

Sorry about the rebase conflicts; this was a particularly fast-moving week! Thanks for working on this!

mishamsk commented on crates/red_knot_python_semantic/src/symbol.rs:
```diff
@@ -183,6 +183,38 @@ pub(crate) fn symbol<'db>(db: &'db dyn Db, scope: ScopeId<'db>, name: &str) -> S
+            }
+
+            if let SymbolAndQualifiers(Symbol::Type(ty, _), _) = symbol_and_quals {
+                // Otherwise, we need to check if the symbol has bindings
+                let use_def = use_def_map(db, scope);
+                let bindings = use_def.public_bindings(symbol);
+                let inferred =
+                    symbol_from_bindings_impl(db, bindings, RequiresExplicitReExport::No);
+
+                match inferred {
+                    Symbol::Unbound => Symbol::Unbound,
+                    Symbol::Type(_, boundness) => Symbol::Type(ty, boundness),
+                }
```
Would be great to merge this with a TODO and I can take a stab at boundness/declaredness refactor with the team permission.

This PR already has 70+ comments, I am afraid it will become harder and harder to review if we keep it open.

carljm commented:
[CodSpeed](https://codspeed.io/astral-sh/ruff/branches/mishamsk%3A15963-fix-diagnostic-for-pure-instance-variables-access-from-class) says this is a 1% cold and 3% incremental regression. That's improved from 5.5% incremental regression in the prior version. I think the current regression is acceptable and within the reasonable range for the added functionality here. It may be that we get some improvement from resolving the TODO around re-resolving bindings.

sharkdp commented on crates/red_knot_python_semantic/src/symbol.rs:
```diff
@@ -377,92 +412,98 @@ impl<'db> From<Symbol<'db>> for SymbolAndQualifiers<'db> {
-                // can assign a "narrow" type to it even if it is not *declared*. This means, we
-                // do not have to call [`widen_type_for_undeclared_public_symbol`].
-                let is_considered_non_modifiable =
-                    is_final || symbol_table(db, scope).symbol(symbol_id).name() == "__slots__";
```
What happened to the `is_final` check here? Maybe we don't have any tests for this, but I think we currently handled `Final` symbols correctly, by not unioning with `Unknown`, because they can not be assigned to from a different scope.

mishamsk replied:
@sharkdp I removed it because it didn't make sense. It was impossible for final qualifier to be set in the branch where it was used

sharkdp replied:
I think it's okay to remove this for now. I am struggling to write a meaningful test for this. Let's just re-introduce this when we support final. I'm also removing the `SymbolAndQualifiers::is_final` method in the meantime.

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/attributes.md:
```diff
@@ -90,13 +89,10 @@ c_instance = C()
 
 reveal_type(c_instance.declared_and_bound)  # revealed: str | None
 
-# TODO: we currently plan to emit a diagnostic here. Note that both mypy
-# and pyright show no error in this case! So we may reconsider this in
-# the future, if it turns out to produce too many false positives.
```
it would be useful to retain a comment here that mypy and pyright do not show an error in this case, but we deliberately deviate from them

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/attributes.md:
```diff
@@ -90,13 +89,10 @@ c_instance = C()
+# error: [unresolved-attribute] "Attribute `declared_and_bound` can only be accessed on instances, not on the class object `Literal[C]` itself."
+reveal_type(C.declared_and_bound)  # revealed: Unknown
 
-# TODO: same as above. We plan to emit a diagnostic here, even if both mypy
-# and pyright allow this.
```
same here, it would be good to keep some version of this comment even if it's no longer a TODO

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/attributes.md:
```diff
@@ -116,11 +112,10 @@ c_instance = C()
 
 reveal_type(c_instance.only_declared)  # revealed: str
 
-# TODO: mypy and pyright do not show an error here, but we plan to emit a diagnostic.
-# The type could be changed to 'Unknown' if we decide to emit an error?
```
and here

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3706,16 +3707,44 @@ impl<'db> TypeInferenceBuilder<'db> {
                 LookupError::Unbound => {
+                    let bound_on_instance = match value_type {
+                    Type::ClassLiteral(ClassLiteralType { class }) => {
+                        !class.instance_member(self.db(), attr).0.is_unbound()
```
there's a `let db = self.db();` assignment a few lines higher so that this method doesn't have to have lots of verbose `self.db()` calls everywhere

```suggestion
                        !class.instance_member(db, attr).0.is_unbound()
```

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3706,16 +3707,44 @@ impl<'db> TypeInferenceBuilder<'db> {
+                    Type::SubclassOf(subclass_of @ SubclassOfType { .. }) => {
+                        match subclass_of.subclass_of() {
+                            ClassBase::Class(class) => {
+                                !class.instance_member(self.db(), attr).0.is_unbound()
```
```suggestion
                                !class.instance_member(db, attr).0.is_unbound()
```

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3706,16 +3707,44 @@ impl<'db> TypeInferenceBuilder<'db> {
+                            ClassBase::Class(class) => {
+                                !class.instance_member(self.db(), attr).0.is_unbound()
+                            }
+                            ClassBase::Dynamic(_) => unreachable!("For any symbol, lookup on a dynamic type will return a bound symbol of the same type"),
```
a couple of nits:

- Lookup on a dynamic type won't always return a bound symbol -- e.g. `tuple[Any, Any]` is a dynamic type, but looking up the attribute "foo" on `tuple[Any, Any]` won't return a bound symbol. This is specifically true for dynamic `SubclassOf` types, but not true in general for _all_ dynamic types.
- It won't necessarily return a bound symbol of the same _type_. That might be what we have right now, but in the long term, for example, `str & Any` might be a better answer for the type of the `__module__` attribute on an object with type `type[Any]`.

```suggestion
                            ClassBase::Dynamic(_) => unreachable!("Attribute lookup on a dynamic `SubclassOf` type should always return a bound symbol"),
```

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3706,16 +3707,44 @@ impl<'db> TypeInferenceBuilder<'db> {
+
+                if bound_on_instance {
+                    self.context.report_lint(
+                            &UNRESOLVED_ATTRIBUTE,
+                            attribute,
+                            format_args!(
+                                "Attribute `{}` can only be accessed on instances, not on the class object `{}` itself.",
+                                attr.id,
+                                value_type.display(self.db())
+                            ),
+                        );
```
some overindentation here

```suggestion
                        &UNRESOLVED_ATTRIBUTE,
                        attribute,
                        format_args!(
                            "Attribute `{}` can only be accessed on instances, not on the class object `{}` itself.",
                            attr.id,
                            value_type.display(db)
                        ),
                    );
```

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3706,16 +3707,44 @@ impl<'db> TypeInferenceBuilder<'db> {
                         format_args!(
                             "Type `{}` has no attribute `{}`",
-                            value_type.display(db),
+                            value_type.display(self.db()),
```
spurious change

```suggestion
                            value_type.display(db),
```

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3745,22 +3774,54 @@ impl<'db> TypeInferenceBuilder<'db> {
+                        let instance_member = instance.class.instance_member(self.db(), attr);
+                        if instance_member.is_class_var() {
+                            self.context.report_lint(
                             &INVALID_ATTRIBUTE_ACCESS,
                             attribute,
                             format_args!(
                                 "Cannot assign to ClassVar `{attr}` from an instance of type `{ty}`",
                                 ty = value_ty.display(self.db()),
                             ),
                         );
```
some underindentation here

```suggestion
                                &INVALID_ATTRIBUTE_ACCESS,
                                attribute,
                                format_args!(
                                    "Cannot assign to ClassVar `{attr}` from an instance of type `{ty}`",
                                    ty = value_ty.display(self.db()),
                                ),
                            );
```

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3745,22 +3774,54 @@ impl<'db> TypeInferenceBuilder<'db> {
+                                Type::SubclassOf(subclass_of @ SubclassOfType { .. }) => {
+                                    match subclass_of.subclass_of() {
+                                        ClassBase::Class(class) => Some(class),
+                                        ClassBase::Dynamic(_) => unreachable!("For any symbol, lookup on a dynamic type will return a bound symbol of the same type"),
```
```suggestion
                                        ClassBase::Dynamic(_) => unreachable!("Attribute lookup on a dynamic `SubclassOf` type should always return a bound symbol"),
```

AlexWaygood reviewed (commented):
thanks again, and sorry we keep creating merge conflicts for you here!

sharkdp commented:
@mishamsk  I can look into making the required changes to get this merged — thank you very much for your work!

mishamsk commented:
> @mishamsk I can look into making the required changes to get this merged — thank you very much for your work!

thanks. I was going to say that I am happy to apply the requested changes myself, but looks like it is too late ;-)

sharkdp reviewed (approved):
Thank you very much. The final benchmarks show a -1% regression on the incremental setup.

mishamsk commented:
@sharkdp thanks for pushing this over the edge. YaY
